src/routes/blog/post/state-of-natural-language-processing/+page.markdoc

```diff
@@ -10,3 +10,152 @@ category: product
++-----+------------+-----+
+      |            |
+      |  car       |
+      |     bike   |
+      +------------+
+```
+
+In this space, similar words like "cat" and "dog" are closer together, while unrelated words like "car" and "bike" are farther apart. This is a simplified view, but it illustrates how word embeddings capture semantic relationships.
+
+The introduction of transformer models in 2017, such as BERT and GPT, using [attention mechanisms](https://arxiv.org/abs/1706.03762), was another major milestone. These models could handle long pieces of text more effectively, leading to better performance on a wide range of NLP tasks.
+
+# Traditional NLP models and embeddings
+
+Before diving into the intricacies of LLMs, let's take a step back and look at some traditional NLP techniques that paved the way for these advanced models.
+
+## Naive Bayes classifiers
+
+Consider building a spam email detector. You could use a Naive Bayes classifier, which learns from examples of spam and non-spam emails to predict whether a new email is spam. 
+
+Here's a simple analogy: 
+- Spam words (e.g., "free", "winner", "viagra") are like red flags 🚩
+- Non-spam words (e.g., "meeting", "project", "dinner") are like green flags 🚩
+
+The Naive Bayes classifier counts the red and green flags in an email and makes a prediction based on which color dominates. It's a straightforward yet effective approach for many text classification tasks.
+
+## Bag of Words (BoW) and TF-IDF
+
+Next up, we have the Bag of Words (BoW) model. Think of it like a grocery bag filled with words from a document. The BoW model cares about which words are present and how many times they appear, but it doesn't care about the order.
+
+```
+Document: "The quick brown fox jumps over the lazy dog"
+BoW: {"the": 2, "quick": 1, "brown": 1, "fox": 1, "jumps": 1, "over": 1, "lazy": 1, "dog": 1}
+```
+
+TF-IDF (Term Frequency-Inverse Document Frequency) takes this a step further by considering how important a word is in a document compared to the entire corpus. It's like giving more weight to the unique items in your grocery bag.
+
+## From rule-based to machine learning approaches
+
+The shift from rule-based systems to machine learning marked a turning point in NLP. Here are a few key developments:
+
+**Long Short-Term Memory (LSTM) Networks**: LSTMs are like the elephants of the neural network world - they have a long memory and can remember important information over long sequences of text.
+
+**Transformers and the Attention Mechanism**: Transformers are the superheroes of NLP. They can process text in parallel (like reading multiple books at once) and use attention to focus on the most important parts (like highlighting key passages).
+
+**Advanced Embeddings (BERT and GloVe)**: These embeddings are like high-resolution maps of language. They capture fine-grained relationships between words and adapt to different contexts, helping models navigate the complex landscape of human language.
```

Comment:
Can we number this list? So there is an indentation in the rendered output ?

Reply:
done